clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/KafkaBasedCommandSender.java

```diff
@@ -124,8 +187,78 @@ public KafkaBasedCommandSender(
         Objects.requireNonNull(deviceId);
         Objects.requireNonNull(command);
 
-        // TODO to implement.
-        return null;
+        final Span span = TracingHelper
+                .buildChildSpan(tracer, context, "send command and wait for response", getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .withTag(TracingHelper.TAG_TENANT_ID, tenantId)
+                .withTag(TracingHelper.TAG_DEVICE_ID, deviceId)
+                .start();
+        final String correlationId = correlationIdSupplier.get();
+        final Promise<Void> requestTracker = Promise.promise();
+        final Promise<DownstreamMessage<KafkaMessageContext>> responseTracker = Promise.promise();
+        final Promise<Void> timeOutTracker = Promise.promise();
+        final long timerId = vertx.setTimer(COMMAND_TIMEOUT, id -> {
+            final SendMessageTimeoutException timeoutException = new SendMessageTimeoutException(
+                    "send command/wait for response timed out after " + COMMAND_TIMEOUT + "ms");
+            //remove the correlation id if timed out
+            Optional.ofNullable(commandResponseSubscriptions.get(tenantId))
+                    .map(Pair::two)
+                    .map(ids -> ids.remove(correlationId));
+            TracingHelper.logError(span, timeoutException);
+            timeOutTracker.fail(timeoutException);
+            requestTracker.tryComplete();
+            responseTracker.tryComplete();
+        });
+        final Handler<DownstreamMessage<KafkaMessageContext>> messageHandler = m -> {
```

Comment:
The `messageHandler` here uses variables specific to this `sendCommand` invocation (`responseTracker`, `timerId`, etc.) but the `messageHandler` is added to the `KafkaBasedDownstreamMessageConsumer` which is only created *once* per tenant.
So it looks to me this doesn't work for any `sendCommand` invocations where `commandResponseSubscriptions.get(tenantId) != null`.

I think `timerId`, `timeOutTracker` and `responseTracker` need to be stored alongside the `correlationId`.

Reply:
@calohmn I have replaced it with a wrapper for a promise with expiring mechanism and store them in a map. Would you mind taking a look?